src/apps/workspaces/service/user_access.py

```diff
@@ -90,6 +97,95 @@ async def get_workspace_applets_by_language(
+        await self._validate_ownership(schema.applet_ids)
+
+        # check if schema.user_id is manager of all applets
+        await self._validate_access(
+            user_id=schema.user_id,
+            removing_applets=schema.applet_ids,
+            roles=[
+                Role.MANAGER,
+                Role.COORDINATOR,
+                Role.EDITOR,
+                Role.REVIEWER,
+            ],
+        )
+
+        # remove manager access
+        for applet_id in schema.applet_ids:
+            await UserAppletAccessCRUD().delete_all_by_user_and_applet(
+                user_id=schema.user_id, applet_id=applet_id
+            )
+
+    async def remove_respondent_access(self, schema: RemoveRespondentAccess):
+        """Remove respondent access from a specific user."""
+        # check if user is owner of all applets
+        await self._validate_ownership(schema.applet_ids)
+
+        # check if schema.user_id is respondent of all applets
+        await self._validate_access(
+            user_id=schema.user_id,
+            removing_applets=schema.applet_ids,
+            roles=[Role.RESPONDENT],
+        )
+
+        # remove respondent access
+        for applet_id in schema.applet_ids:
+            await UserAppletAccessCRUD().delete_all_by_user_and_applet(
+                user_id=schema.user_id, applet_id=applet_id
+            )
+
+        # delete all responses of respondent in applets
+        if schema.delete_responses:
+            for applet_id in schema.applet_ids:
+                await AnswerActivityItemsCRUD().delete_by_applet_user(
+                    applet_id=applet_id, user_id=schema.user_id
+                )
+                await AnswerFlowItemsCRUD().delete_by_applet_user(
+                    user_id=schema.user_id, applet_id=applet_id
+                )
+
+    async def _validate_ownership(self, applet_ids: list[uuid.UUID]):
+        owners_applet_ids = [
+            owner_applet.applet_id
+            for owner_applet in (
+                await UserAppletAccessCRUD().get_all_by_user_id_and_roles(
+                    self._user_id, roles=[Role.ADMIN]
+                )
+            )
+        ]
+        for applet_id in applet_ids:
+            if applet_id not in owners_applet_ids:
+                raise AppletAccessDenied(
+                    message=f"User is not owner of applet {applet_id}"
+                )
+
+    async def _validate_access(
+        self,
+        user_id: uuid.UUID,
+        removing_applets: list[uuid.UUID],
+        roles: list[Role],
+    ):
+        # check if user_id has access to applets with roles
+        applet_ids = [
+            manager_applet.applet_id
+            for manager_applet in (
+                await UserAppletAccessCRUD().get_all_by_user_id_and_roles(
+                    user_id,
+                    roles=roles,
+                )
+            )
+        ]
+
+        for applet_id in removing_applets:
+            if applet_id not in applet_ids:
```

Comment:
pls do not iterate, check by length like `len(removing_applets) != len(applet_ids)` where each part is list of unique ids

Reply:
or use `set` to iterate and search item inside list if you need exact access error